File: src/ui.js
export const ESC = '\x1b[';
export const colors = {
    reset: `${ESC}0m`,
    bold: `${ESC}1m`,
    dim: `${ESC}2m`,
    magenta: `${ESC}35m`,
    blue: `${ESC}34m`,
    cyan: `${ESC}36m`,
    green: `${ESC}32m`,
    yellow: `${ESC}33m`,
    red: `${ESC}31m`,
};

export const ui = {
    log: (msg) => console.log(msg),
    error: (msg) => console.log(`${colors.red}✖ ${msg}${colors.reset}`),
    warn: (msg) => console.log(`${colors.yellow}⚠️  ${msg}${colors.reset}`),
    success: (msg) => console.log(`${colors.green}✔ ${msg}${colors.reset}`),
    info: (msg) => console.log(`${colors.cyan}ℹ ${msg}${colors.reset}`),
    clearLine: () => process.stdout.write('\r\x1b[K'),
};

export class Spinner {
    constructor(text) {
        this.text = text;
        this.frames = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
        this.index = 0;
        this.timer = null;
    }
    start() {
        this.timer = setInterval(() => {
            process.stdout.write(`\r${colors.cyan}${this.frames[this.index]}${colors.reset} ${this.text}`);
            this.index = (this.index + 1) % this.frames.length;
        }, 80);
        return this;
    }
    update(text) {
        this.text = text;
    }
    stop(symbol = ' ', color = colors.reset) {
        clearInterval(this.timer);
        ui.clearLine();
    }
    succeed(text) {
        this.stop();
        console.log(`${colors.green}✔${colors.reset} ${text}`);
    }
    fail(text) {
        this.stop();
        console.log(`${colors.red}✖${colors.reset} ${text}`);
    }
}

File: src/ui.rs
use colored::*;
use indicatif::{ProgressBar, ProgressStyle};
use std::time::Duration;

pub struct Spinner {
    pb: ProgressBar,
}

impl Spinner {
    pub fn new(msg: &str) -> Self {
        let pb = ProgressBar::new_spinner();
        pb.enable_steady_tick(Duration::from_millis(120));
        pb.set_style(
            ProgressStyle::with_template("{spinner:.blue} {msg}")
                .unwrap()
                .tick_strings(&["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"]),
        );
        pb.set_message(msg.to_string());
        Self { pb }
    }

    pub fn update(&self, msg: &str) {
        self.pb.set_message(msg.to_string());
    }

    pub fn succeed(&self, msg: &str) {
        self.pb.finish_with_message(format!("{} {}", "✔".green(), msg));
    }

    pub fn fail(&self, msg: &str) {
        self.pb.finish_with_message(format!("{} {}", "✖".red(), msg));
    }

    pub fn stop(&self) {
        self.pb.finish_and_clear();
    }
}

pub fn info(msg: &str) {
    println!("{} {}", "ℹ".blue(), msg);
}

pub fn warn(msg: &str) {
    println!("{} {}", "⚠".yellow(), msg);
}

pub fn error(msg: &str) {
    eprintln!("{} {}", "✖".red(), msg);
}

